entity-service-impl/src/main/java/org/hypertrace/entity/data/service/EntityDataServiceImpl.java

```diff
@@ -496,6 +500,58 @@ public void getEnrichedEntityByTypeAndIdentifyingProps(
         responseObserver);
   }
 
+  @Override
+  public void mergeAndUpsertEntity(
+      MergeAndUpsertEntityRequest request,
+      StreamObserver<MergeAndUpsertEntityResponse> responseObserver) {
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    String tenantId = requestContext.getTenantId().orElse(null);
+    if (tenantId == null) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+
+    Entity receivedEntity = this.entityNormalizer.normalize(tenantId, request.getEntity());
+    Optional<Entity> existingEntity =
```

Comment:
Can't we leverge https://github.com/hypertrace/document-store/blob/081cf1b5f9433e22d1a8736d01e1cc85dd663109/document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java#L114 here?

Reply:
This api doesn't support merging - here we need to fetch any existing entity. If none, create. If exists, evaluate condition and if condition succeeds, upsert. I mentioned the scenario on that PR, but dropped it since it seemed a little specific for ES.